boot/BootFromDevice.c

```diff
@@ -0,0 +1,364 @@
+	CONFIGENTRY* config = NULL;
+	CONFIGENTRY* cfgLinux;
+	char* szGrub;
+
+	szGrub = InitGrubRequest(GRUB_REQUEST_SIZE, drive, partition);
+	cfgLinux = DetectLinuxNative(szGrub);
+	if (cfgLinux != NULL)
+	{
+		FillConfigEntries(cfgLinux, BOOT_NATIVE, drive, partition);
+		config = AddNestedConfigEntry(config, cfgLinux, "Linux");
+	}
+	free(szGrub);
+
+	return config;
+}
+
+int BootFromNative(CONFIGENTRY *config) {
+	char* szGrub;
+	int result = 0;
+
+	DVDTrayClose();
+
+	szGrub = InitGrubRequest(GRUB_REQUEST_SIZE, config->drive, config->partition);
+	if (config->bootSystem == SYS_LINUX)
+		result = LoadLinuxNative(szGrub, config);
+	free(szGrub);
+
+	return result;
+}
+
+CONFIGENTRY* DetectSystemFatX(void) {
+	CONFIGENTRY* config = NULL;
+	CONFIGENTRY* cfgLinux;
+	FATXPartition* partition;
+
+	partition = OpenFATXPartition(0, SECTOR_STORE, STORE_SIZE);
+	if (partition == NULL) return NULL;
+
+	cfgLinux = DetectLinuxFATX(partition);
+	if (cfgLinux != NULL)
+	{
+		FillConfigEntries(cfgLinux, BOOT_FATX, 0, 0);
+		config = AddNestedConfigEntry(config, cfgLinux, "Linux");
+	}
+	CloseFATXPartition(partition);
+
+	return config;
+}
+
+int BootFromFatX(CONFIGENTRY *config) {
+	FATXPartition* partition;
+	int result = 0;
+
+	DVDTrayClose();
+
+	partition = OpenFATXPartition(0, SECTOR_STORE, STORE_SIZE);
+	if (partition == NULL) return false;
+
+	if (config->bootSystem == SYS_LINUX)
+		result = LoadLinuxFATX(partition, config);
+	CloseFATXPartition(partition);
+
+	return result;
+}
+
+CONFIGENTRY *DetectSystemCD(int cdromId) {
+	int n;
+	CONFIGENTRY* config = NULL;
+	CONFIGENTRY* cfgLinux;
+	int nTempCursorX, nTempCursorY;
+
+	printk("\2Please wait\n\n");
+	//See if we already have a CD in the drive
+	//Try for 8 seconds - takes a while to 'spin up'.
+	nTempCursorX = VIDEO_CURSOR_POSX;
+	nTempCursorY = VIDEO_CURSOR_POSY;
+again:
+	DVDTrayClose();
+	printk("Detecting system on CD... \n");
+	for (n = 0; n < 32; ++n) {
+		cfgLinux = DetectLinuxCD(cdromId);
+		if (cfgLinux != NULL) {
+			break;
+		}
+		wait_ms(250);
+	}
+
+	//We couldn't read the disk, so we eject the drive so the user can insert one.
+	if (cfgLinux == NULL) {
+		printk("Boot from CD failed.\nCheck that supported system exists on the CD.\n\n");
+		//Needs to be changed for non-xbox drives, which don't have an eject line
+		//Need to send ATA eject command.
+		DVDTrayEject();
+		wait_ms(2000); // Wait for DVD to become responsive to inject command
+
+		VIDEO_ATTR = 0xffeeeeff;
+		printk("\2Please insert CD and press Button A\n\n\2Press Button B to return to main menu\n\n");
+
+		while (1) {
+			// Retry system detection
+			cfgLinux = DetectLinuxCD(cdromId);
+
+			// Make button 'A' close the DVD tray
+			if (risefall_xpad_BUTTON(TRIGGER_XPAD_KEY_A) == 1) {
+				DVDTrayClose();
+				wait_ms(500);
+				break;
+			}
+			else if (DVD_TRAY_STATE == DVD_CLOSING) {
+				//It's an xbox drive, and somebody pushed the tray in manually
+				wait_ms(500);
```

Comment:
I don't think it's good idea to try reading from the drive, when it's in the middle state, i.e. opening or closing.

Reply:
While I don't disagree with that, I don't think this check prevents that, there is no wait loop for `DVD_CLOSED`. Older drives may have trouble closing the tray, I can easily imagine a case where it takes >2 seconds to close the drive.

I think this whole flow should be re-factored so it can rely on the tray state better (not having tray state on non-xbox drives might be a blocker though?). But regardless of my thoughts, I think this would be a larger task outside of this PR 👍